hippy/module/pypy_bridge/py_adapters.py

```diff
@@ -363,16 +371,16 @@ def get_wrapped_py_obj(self):
 
     def next(self, space):
         index = self.index
-        w_py_value = self.storage_w[index]
+        w_py_value = self.py_space.getitem(self.w_py_list, self.py_space.wrap(self.index))
         self.index = index + 1
-        self.finished = self.index == len(self.storage_w)
+        self._compute_finished_flag()
         return w_py_value.to_php(self.py_space.get_php_interp())
 
```

Comment:
There is a lot of code duplication between the ``next`` and ``next_item`` methods, which should be removed. The only thing that's really different is the return value. So there could be a helper method ``_next`` (or whatever) that returns a tuple of ``(index, converted_result)`` which is used by both methods. This would also make the somewhat weird ``_compute_finished_flag`` unnecessary.